### PR TITLE
Add is_identity to CoordinateMaps

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

Adds `is_identity` to CoordinateMaps for the future functionality of skipping maps in compositions
when that map is the identity.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by nilsdeppe — COMMENTED** [review]
General observation I have is that there aren't tests for the individual maps to check that the identity works correctly there. We could probably have a generic test where you pass in a valid range in logical coords and make sure you get (approximately) the same points out. What do you think?

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 1]
```diff
@@ -458,8 +467,11 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::inv_jacobian_impl(
             });
 
         if (LIKELY(count != 0)) {
-          mapped_point =
-              std::get<(count != 0 ? count - 1 : 0)>(maps)(mapped_point);
+          const auto& map_in_loop =
+              std::get<(count != 0 ? count - 1 : 0)>(maps);
+          if (not map_in_loop.is_identity()) {
+            mapped_point = map_in_loop(mapped_point);
+          }
```
Comment (wthrowe) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: Skip the matrix multiplication by the identity matrix for identity maps.  (And again for the Jacobian code below.)

File: tests/Unit/Domain/CoordinateMaps/Test_CoordinateMap.cpp [hunk 2]
```diff
@@ -748,6 +758,98 @@ void test_make_vector_coordinate_map_base() {
   CHECK(*translated_lower_xi_wedge_base == *(vector_of_translated_wedges[2]));
   CHECK(*translated_lower_eta_wedge_base == *(vector_of_translated_wedges[3]));
 }
+
+void test_coordinate_maps_are_identity() {
+  CHECK(CoordinateMaps::Affine{-1.0, 1.0, -1.0, 1.0}.is_identity());
```
Comment (wthrowe) [inline tests/Unit/Domain/CoordinateMaps/Test_CoordinateMap.cpp]: Is there a reason these are all grouped together here instead of being in the tests for the maps themselves?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp [hunk 3]
```diff
@@ -63,6 +64,19 @@ void check_if_maps_are_equal(
   }
 }
 
+/// \ingroup TestingFrameworkGroup
+/// \brief Given a coordinate map, check that this map is equal to the identity
+/// by evaluating the map at a random set of points.
+template <typename Map>
+void check_if_map_is_identity(const Map& map) {
+  using IdentityMap = CoordinateMaps::Identity<Map::dim>;
+  const IdentityMap id{};
```
Comment (wthrowe) [inline tests/Unit/Domain/CoordinateMaps/TestMapHelpers.hpp]: Unused.

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 4]
```diff
@@ -458,8 +467,11 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::inv_jacobian_impl(
...
 
         if (LIKELY(count != 0)) {
-          mapped_point =
-              std::get<(count != 0 ? count - 1 : 0)>(maps)(mapped_point);
+          const auto& map_in_loop =
+              std::get<(count != 0 ? count - 1 : 0)>(maps);
+          if (not map_in_loop.is_identity()) {
+            mapped_point = map_in_loop(mapped_point);
+
           inv_jac_overload(
```
Comment (wthrowe) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: Reindent.

**Review by wthrowe — CHANGES_REQUESTED** [review]
Add tests that is_identity() is false for an appropriate case for each map.

**Comment by wthrowe** [comment]
Looks good now.

File: tests/Unit/Domain/CoordinateMaps/Test_Wedge2D.cpp [hunk 5]
```diff
@@ -150,43 +150,34 @@ void test_wedge2d_fail() noexcept {
...
-      CoordinateMaps::Wedge2D(0.2, 4.0, 0.0, 1.0, OrientationMap<2>{}, false);
-  CHECK_FALSE(wedge2d == wedge2d_inner_radius_changed);
-  CHECK_FALSE(wedge2d == wedge2d_outer_radius_changed);
-  CHECK_FALSE(wedge2d == wedge2d_inner_circularity_changed);
-  CHECK_FALSE(wedge2d == wedge2d_outer_circularity_changed);
-  CHECK_FALSE(wedge2d == wedge2d_orientation_map_changed);
-  CHECK_FALSE(wedge2d == wedge2d_use_equiangular_map_changed);
+  test_wedge2d_all_orientations(false);  // Equidistant
+  test_wedge2d_all_orientations(true);   // Equiangular
+  SECTION("Equality") {
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_Wedge2D.cpp]: SECTION reruns the test from the beginning. Is that really what you want here? I think you can just use scope + a comment. You could also put all the blocks into SECTIONs like you've done in other tests.

File: tests/Unit/Domain/CoordinateMaps/Test_SpecialMobius.cpp [hunk 6]
```diff
@@ -14,24 +14,22 @@
...
-SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.SpecialMobius.Suite",
-                  "[Domain][Unit]") {
-  // Set up random number generator
-  MAKE_GENERATOR(gen);
-  // Note: Empirically we have found that the map is accurate
-  // to 12 decimal places for mu = 0.96.
-  // test_suite demands more accuracy than 12 decimal places
-  // so we narrow the range of mu correspondingly.
-  std::uniform_real_distribution<> mu_dis(-0.85, 0.85);
+SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.SpecialMobius", "[Domain][Unit]"){
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_SpecialMobius.cpp]: It looks like the formatting is off in this test

File: tests/Unit/Domain/CoordinateMaps/Test_ProductMaps.cpp [hunk 7]
```diff
@@ -56,290 +56,317 @@ void test_product_two_maps_fail() {
...
+                                       CoordinateMaps::Affine>{
+            CoordinateMaps::Affine{-1.0, 1.0, -1.0, 1.0},
+            CoordinateMaps::Affine{-1.0, 1.0, -1.0, 1.0}});
+    CHECK(not CoordinateMaps::ProductOf2Maps<CoordinateMaps::Affine,
+                                             CoordinateMaps::Affine>{
+        CoordinateMaps::Affine{-1.0, 2.0, -1.0, 1.0},
+        CoordinateMaps::Affine{-1.0, 1.0, -3.0, 1.0}}
+                  .is_identity());
+  }
+  SECTION("Product of 3 Maps") {
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_ProductMaps.cpp]: I think it would be much cleaner if you separated these sections into functions. I personally prefer that style because it keeps functions small and reduces how much scope can spill. For example, you can put two functions `void test_product_2_maps()` and `void product_3_maps()` into an anonymous namespace and call them both from a single `SPECTRE_TEST_CASE`

File: tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp [hunk 8]
```diff
@@ -77,108 +77,108 @@ void test_frustum_fail() noexcept {
...
   test_frustum_fail();
-}
-
-SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.Frustum.Equidistant",
-                  "[Domain][Unit]") {
-  test_suite_for_frustum(false);
-}
+  test_suite_for_frustum(false);  // Equidistant
+  test_suite_for_frustum(true);   // Equiangular
+  SECTION("Alignment") {
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_Frustum.cpp]: Again, just factor into a separate function and call the function from the main test case

File: tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp [hunk 9]
```diff
@@ -15,20 +15,19 @@
 #include "tests/Unit/TestHelpers.hpp"
 
 SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.EquatorialCompression",
-                  "[Domain][Unit]") {
-  // Set up random number generator
-  MAKE_GENERATOR(gen);
-  std::uniform_real_distribution<> aspect_ratio_dis(0.1, 10);
+                  "[Domain][Unit]"){
+    SECTION("Test Suite"){// Set up random number generator
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_EquatorialCompression.cpp]: Again with SECTION -> function in anonymous namespace

File: tests/Unit/Domain/CoordinateMaps/Test_DiscreteRotation.cpp [hunk 10]
```diff
@@ -21,81 +21,99 @@
 
 SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.DiscreteRotation.Specific1D",
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_DiscreteRotation.cpp]: Here too. Instead of sections factor into functions that are called from the SPECTRE_TEST_CASE

File: tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp [hunk 11]
```diff
@@ -57,182 +57,198 @@ void cubic_scale_non_invertible(const double a0, const double b0,
...
-  };
-
-  // test inverse at inner and outer boundary values
-  run_tests(0.0);
-  run_tests(20.0);
+    run_tests(init_func_a2, init_func_b2, 6.0, 5.0, 0.0);
+  }
+
+  // Test boundaries
+  {
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp]: Again, please factor these scopes into separate functions. You should generally avoid functions that don't fit on your full laptop screen vertically. The longer the function the more difficult it is to understand, reason about, and catch bugs in.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
A general comment, this PR went from adding is_identity to a major refactor of the maps tests. In the future please try to keep those separate so it's easier to review.

**Comment by kidder** [comment]
Can you squash this into two or three commits; one or two for is_identity and one for unrelated test code reorganizations

**Comment by kidder** [comment]
you can squash the changes you made addressing @wthrowe 's comments since he indicated he was happy with them.

File: tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp (hunk 11 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/CoordinateMaps/Test_CubicScale.cpp]: ping on this

File: tests/Unit/Domain/CoordinateMaps/Test_DiscreteRotation.cpp [hunk 12]
```diff
@@ -95,12 +97,36 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.DiscreteRotation.Specific1D",
     }
   }
 
-  check_if_map_is_identity(
-      CoordinateMaps::DiscreteRotation<1>{OrientationMap<1>{}});
+  void test_is_identity() {
+    check_if_map_is_identity(
+        CoordinateMaps::DiscreteRotation<1>{OrientationMap<1>{}});
+    CHECK(not CoordinateMaps::DiscreteRotation<1>{
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_DiscreteRotation.cpp]: Please make sure this gets squashed into the right commit :)

**Review by nilsdeppe — COMMENTED** [review]
Just need to separate out `Test_CubicScale` to have 2 functions instead of the scope. We're happy for you to squash things into the appropriate commits. Review with @kidder @wthrowe

File: tests/Unit/Domain/CoordinateMaps/Test_DiscreteRotation.cpp [hunk 13]
```diff
@@ -37,65 +37,71 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.DiscreteRotation.Specific1D",
...
   CHECK(rotation_nx(point_px) == point_nx);
-}
-
-SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.DiscreteRotation.Specific2D",
-                  "[Domain][Unit]") {
-  const CoordinateMaps::DiscreteRotation<2> identity_map2d{};
-  const CoordinateMaps::DiscreteRotation<2> rotation_ny_px{
-      OrientationMap<2>{std::array<Direction<2>, 2>{
-          {Direction<2>::lower_eta(), Direction<2>::upper_xi()}}}};
+  }
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_DiscreteRotation.cpp]: this line is causing the goofy indentation below :)

File: tests/Unit/Domain/CoordinateMaps/Test_Rotation.cpp [hunk 14]
```diff
@@ -65,79 +108,41 @@ SPECTRE_TEST_CASE("Unit.Domain.CoordinateMaps.Rotation<2>", "[Domain][Unit]") {
...
-          approx(gsl::at(eta, i)));
-    CHECK(gsl::at(three_dim_rotation_map.inverse(zeta_hat).get(), i) ==
-          approx(gsl::at(zeta, i)));
-    CHECK(inv_jac.get(0, i) == approx(gsl::at(xi_hat, i)));
-    CHECK(inv_jac.get(1, i) == approx(gsl::at(eta_hat, i)));
-    CHECK(inv_jac.get(2, i) == approx(gsl::at(zeta_hat, i)));
-    CHECK(jac.get(i, 0) == approx(gsl::at(xi_hat, i)));
-    CHECK(jac.get(i, 1) == approx(gsl::at(eta_hat, i)));
-    CHECK(jac.get(i, 2) == approx(gsl::at(zeta_hat, i)));
   }
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_Rotation.cpp]: here it's this line

File: tests/Unit/Domain/CoordinateMaps/Test_Rotation.cpp [hunk 15]
```diff
@@ -108,41 +109,46 @@ void test_rotation_2() {
...
-                    std::array<double, 3>{{-1.0, 0.0, 0.0}});
-
-    test_rotation_3(CoordinateMaps::Rotation<3>(0.0, M_PI_2, 0.0),
-                    std::array<double, 3>{{0.0, 0.0, -1.0}},
-                    std::array<double, 3>{{0.0, 1.0, 0.0}},
-                    std::array<double, 3>{{1.0, 0.0, 0.0}});
-  }
+  check_if_map_is_identity(CoordinateMaps::Rotation<2>{0.0});
+  CHECK(not CoordinateMaps::Rotation<2>(M_PI_2).is_identity());
+}
```
Comment (nilsdeppe) [inline tests/Unit/Domain/CoordinateMaps/Test_Rotation.cpp:116]: It looks like you got the formatting restored in this commit. It'll be a lot easier to use `git blame` if you make sure this formatting is correct in the first commit.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
You can probably squash the commits `Add check_if_map_is_identity` and `Add is_identity to CoordinateMaps` into one, but please wait for @wthrowe to let us know whether or not he is okay with that.

**Comment by wthrowe** [comment]
You should get the formatting changes into the correct commits.  Merging the two identity-related commits sounds good, but I don't have a strong preference.

**Review by nilsdeppe — DISMISSED** [review]
LGTM. @wthrowe are you happy with this?

File: src/Domain/CoordinateMaps/ProductMaps.hpp [hunk 16]
```diff
@@ -242,22 +248,29 @@ class ProductOf3Maps {
...
                          const ProductOf3Maps& rhs) noexcept {
     return lhs.map1_ == rhs.map1_ and lhs.map2_ == rhs.map2_ and
-           lhs.map3_ == rhs.map3_;
+           lhs.map3_ == rhs.map3_ and lhs.is_identity_ == rhs.is_identity_;
   }
 
   Map1 map1_;
   Map2 map2_;
   Map3 map3_;
+  bool is_identity_;
```
Comment (kidder) [inline src/Domain/CoordinateMaps/ProductMaps.hpp]: default to false

File: src/Domain/CoordinateMaps/ProductMaps.hpp [hunk 17]
```diff
@@ -131,19 +131,25 @@ class ProductOf2Maps {
...
   friend bool operator==(const ProductOf2Maps& lhs,
                          const ProductOf2Maps& rhs) noexcept {
-    return lhs.map1_ == rhs.map1_ and lhs.map2_ == rhs.map2_;
+    return lhs.map1_ == rhs.map1_ and lhs.map2_ == rhs.map2_ and
+           lhs.is_identity_ == rhs.is_identity_;
   }
 
   Map1 map1_;
   Map2 map2_;
+  bool is_identity_;
```
Comment (kidder) [inline src/Domain/CoordinateMaps/ProductMaps.hpp]: default to false

File: src/Domain/CoordinateMaps/DiscreteRotation.hpp [hunk 18]
```diff
@@ -58,13 +58,17 @@ class DiscreteRotation {
...
  private:
   friend bool operator==(const DiscreteRotation& lhs,
                          const DiscreteRotation& rhs) noexcept {
-    return lhs.orientation_ == rhs.orientation_;
+    return lhs.orientation_ == rhs.orientation_ and
+           lhs.is_identity_ == rhs.is_identity_;
   }
 
   OrientationMap<VolumeDim> orientation_{};
+  bool is_identity_;
```
Comment (kidder) [inline src/Domain/CoordinateMaps/DiscreteRotation.hpp]: default to false

File: src/Domain/CoordinateMaps/BulgedCube.hpp [hunk 19]
```diff
@@ -513,14 +513,18 @@ class BulgedCube {
...
  private:
   template <typename T>
   std::array<tt::remove_cvref_wrap_t<T>, 3> xi_derivative(
       const std::array<T, 3>& source_coords) const noexcept;
   friend bool operator==(const BulgedCube& lhs, const BulgedCube& rhs) noexcept;
+
   double radius_;
   double sphericity_;
   bool use_equiangular_map_;
+  bool is_identity_;
```
Comment (kidder) [inline src/Domain/CoordinateMaps/BulgedCube.hpp]: default to false

**Review by kidder — COMMENTED** [review]
you can squash these changes

**Review by nilsdeppe — DISMISSED** [review]
@kidder @wthrowe could you please give feedback on this PR when you get a chance? Thanks! :)

**Review by kidder — DISMISSED** [review]

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 20]
```diff
@@ -385,7 +388,9 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::inverse_impl(
          /*f_of_ts*/,
          const std::false_type /*is_time_independent*/) noexcept {
         if (point) {
-          point = the_map.inverse(point.get());
+          if (not the_map.is_identity()) {
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: this should be marked as `LIKELY`

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 21]
```diff
@@ -353,7 +354,9 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::call_impl(
          const std::unordered_map<std::string, FunctionOfTime&>&
          /*f_of_ts*/,
          const std::false_type /*is_time_independent*/) noexcept {
-        point = the_map(point);
+        if (not the_map.is_identity()) {
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: Should be marked as `LIKELY`

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 22]
```diff
@@ -458,24 +467,27 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::inv_jacobian_impl(
...
-          for (size_t source = 0; source < dim; ++source) {
-            for (size_t target = 0; target < dim; ++target) {
-              gsl::at(temp, target) =
-                  inv_jac.get(source, 0) * temp_inv_jac.get(0, target);
-              for (size_t dummy = 1; dummy < dim; ++dummy) {
-                gsl::at(temp, target) += inv_jac.get(source, dummy) *
-                                         temp_inv_jac.get(dummy, target);
+          const auto& map_in_loop =
+              std::get<(count != 0 ? count - 1 : 0)>(maps);
+          if (not map_in_loop.is_identity()) {
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: Should be marked as `LIKELY`

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 23]
```diff
@@ -522,7 +534,11 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::jacobian_impl(
                const std::unordered_map<std::string, FunctionOfTime&>&
                /*f_of_ts*/,
                const std::false_type /*is_time_independent*/) {
-              *no_frame_jac = the_map.jacobian(point);
+              if (not the_map.is_identity()) {
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: should be marked as `LIKELY`

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 24]
```diff
@@ -536,24 +552,26 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::jacobian_impl(
...
-          for (size_t source = 0; source < dim; ++source) {
-            for (size_t target = 0; target < dim; ++target) {
-              gsl::at(temp, target) =
-                  noframe_jac.get(target, 0) * jac.get(0, source);
-              for (size_t dummy = 1; dummy < dim; ++dummy) {
-                gsl::at(temp, target) +=
-                    noframe_jac.get(target, dummy) * jac.get(dummy, source);
+          const auto& map_in_loop =
+              std::get<(count != 0 ? count - 1 : 0)>(maps);
+          if (not map_in_loop.is_identity()) {
```
Comment (nilsdeppe) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: should be marked as `LIKELY`

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by mar-celine** [comment]
Force pushed 2/14:
* Nils' comments on `LIKELY`
* IWYU in BulgedCube.hpp (include `limits`)

File: src/Domain/CoordinateMaps/CoordinateMap.hpp [hunk 25]
```diff
@@ -445,7 +450,11 @@ CoordinateMap<SourceFrame, TargetFrame, Maps...>::inv_jacobian_impl(
                const std::unordered_map<std::string, FunctionOfTime&>&
                /*f_of_ts*/,
                const std::false_type /*is_time_independent*/) {
-              *t_inv_jac = the_map.inv_jacobian(point);
+              if (not the_map.is_identity()) {
```
Comment (kidder) [inline src/Domain/CoordinateMaps/CoordinateMap.hpp]: why not this one? (as LIKELY)

**Comment by mar-celine** [comment]
Force push 2/14 13:00:
Larry's `LIKELY`

**Review by kidder — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]

**Comment by kidder** [comment]
@wthrowe please look at (and merge if you are happy)

**Review by wthrowe — DISMISSED** [review]
Still waiting on Travis.  I'll check back later to merge (unless someone else gets to it first).

**Comment by nilsdeppe** [comment]
This needs a rebase. It's not clear to me all the test failures are because of the GCC segfaults.

**Comment by mar-celine** [comment]
Force push 2/15 10:56: rebase on develop

**Review by nilsdeppe — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]
Looks like just the usual failures now.

**Review by kidder — APPROVED** [review]
